asf_heat_pump_suitability/pipeline/prepare_features/garden_size.py

```diff
@@ -0,0 +1,125 @@
+
+    Args:
+        land_files_gdf (gpd.GeoDataFrame): INSPIRE land file names and file bounding polygons
+        building_files_gdf (gpd.GeoDataFrame): Microsoft building footprint file names and file bounding polygons
+
+    Returns:
+        Mapping: mapping where keys are INSPIRE land parcel file names and values are Microsoft building footprint file
+        names
+    """
+    logging.info("Mapping building footprint files to INSPIRE land registry files")
+    gdf = gpd.overlay(land_files_gdf, building_files_gdf, how="intersection")
+    gdf = gdf.sort_values(by="inspire_file_name")
+    file_dict = pd.Series(
+        gdf["ms_url"].values, index=gdf["inspire_file_name"]
+    ).to_dict()
+
+    return file_dict
+
+
+def generate_gdf_land_building_overlay(
+    land_parcels_gdf: gpd.GeoDataFrame,
+    building_footprints_gdf: gpd.GeoDataFrame,
+    outbuilding_size: int = 30,
+    s_building_prop: float = 0.45,
+    l_building_prop: float = 0.05,
+) -> gpd.GeoDataFrame:
+    """
+    Generate intersections between land parcel polygons and building footprint polygons, with rules applied to remove
+    suspected erroneous building intersections (i.e. small corners or slivers of buildings) but retain outbuildings.
+    Building intersections are kept if they meet the following conditions:
+    1. building intersection area <= outbuilding_size and >= s_building_prop * outbuilding_size
+    2. building intersection area > outbuilding_size and >= l_building_prop * outbuilding_size
+
+    Args:
+        land_parcels_gdf (gpd.GeoDataFrame): land parcel polygons
+        building_footprints_gdf (gpd.GeoDataFrame): building footprint polygons
+        outbuilding_size (int): max area (m2) of building footprints assumed to be outbuildings. Default 30m2.
+        s_building_prop (float): minimum proportion of buildings smaller than outbuilding_size. Default 45%.
+        l_building_prop (float): minimum proportion of buildings larger than outbuilding_size. Default 5%.
+
+    Returns:
+        gpd.GeoDataFrame: intersections of land parcel polygons and building footprint polygons
+    """
+    gdf = gpd.overlay(land_parcels_gdf, building_footprints_gdf, how="intersection")
+    gdf["building_intersection_area_m2"] = gdf["geometry"].area
+
+    gdf["min_size_of_small_building"] = gdf["building_area_m2"] * s_building_prop
+    gdf["min_size_of_large_building"] = gdf["building_area_m2"] * l_building_prop
+
+    gdf = gdf[
+        (
+            (gdf["building_intersection_area_m2"] <= outbuilding_size)
+            & (
+                gdf["building_intersection_area_m2"]
+                >= gdf["min_size_of_small_building"]
+            )
+        )
+        | (
+            (gdf["building_intersection_area_m2"] > outbuilding_size)
+            & (
+                gdf["building_intersection_area_m2"]
+                >= gdf["min_size_of_large_building"]
+            )
+        )
```

Comment:
I'm not 100% confident in this logic as I find it quite confusing to make sense of.

To be included as a valid intersection between a building polygon and land parcel, the following conditions must be met, either:

The candidate intersection is less than or equal to the maximum out building size and greater than or equal to 45% of the original building size.

or,

The candidate intersection is larger than the maximum size of an outbuilding and greater than or equal to 5% of the original building size.

I think this makes sense, but I think it'd be worthwhile revisiting once you've adjusted the building footprint coordinate conversions to make sure it's still performing as you expect.